File: Client/src/modules/models/spell/PowerSpell.js
var PowerSpell = Spell.extend({
    ctor:function(){
        //Spell.prototype.ctor(3).call();
        //this.init(5);
        this._super(5);
        this.effectTick = 15 * GC.BATTLE.TICK_PER_SECOND / 10;
        this.range = 0.6;
    },
    setUI:function (spellUI){
        this.ui = spellUI;
    },
    update:function (monsters, towers){
        //cc.log("SpeedUpSpell update:" + this.leftTick);
        if(this.leftTick > 0){
            for(let i = 0; i < towers.length; ++i){
                let tower = towers[i];
                if (Helper.getDistance(this.curPos, tower.logicPoint) <= this.range && tower.isBuffBySpell(this) == false) {
                    tower.addSpell(this);
                    tower.increaseDamage(this.battleId, 1.5);
                }
            }
            this.leftTick -= 1;
            return true;
        }
        else {
            if(this.ui){
                this.ui.removeFromParent();
            }

            return false;
        }
    },
    removeEffect:function (object){
        object.removeDamage(this.battleId);
    },
    doAction:function (object){

    },
    setBattleId:function (battleId){
        this.battleId = battleId;
    },
    doEffect:function (tower){
        tower.increaseDamage(this.battleId, 1.5);
    },
    initAnim:function (){
        let path = this.animPath;
        this.anim = new sp.SkeletonAnimation(path + ".json", path + ".atlas");
        this.anim.setAnchorPoint(cc.p(0.5, 0.5));
        this.anim.setPosition(cc.p(0, 0));
        this.anim.setAnimation(0, this.animArr[0], true);
        this.anim.update(0.0);
        return this.anim;
    }
});

File: Server/src/model/Spell/PowerSpell.java
package model.Spell;

import model.Battle.BattleObject;
import model.Define.ModelDefine;
import model.Monster.Monster;
import model.Tower.Tower;

import java.util.Hashtable;
import java.util.Iterator;
import java.util.LinkedList;

public class PowerSpell extends Spell{
    private Hashtable<Monster, Double> monsters;
    private LinkedList<Monster> targets;
    public PowerSpell(int level) {
        super(5, 1);
        readSpellJson();
        targets = new LinkedList<>();
    }
    protected void readSpellJson(){
        super.readSpellJson();
        this.leftTick = (int)this.duration * ModelDefine.TICK_PER_SECOND;
        this.effectTick = 15 * ModelDefine.TICK_PER_SECOND / 10;
    }
    public boolean update(LinkedList<Monster>monsters, LinkedList<Tower>towers){
        System.out.println("PowerSpell " + battleObjectId + " time left: " + leftTick);
        if(leftTick > 0){
            Iterator<Tower> itr = towers.iterator();
            while(itr.hasNext()) {
                Tower tower = itr.next();
                if (curPos.getDistance(tower.getCurPos()) <= range && tower.isBuffBySpell(this) == false) {
                    tower.addSpell(this);
                    tower.increaseDamage(battleObjectId,1.5);
                }
            }
            leftTick -= 1;
            return true;
        }
        else {
            return false;
        }
    }
    public void doAction(BattleObject battleObject){

    }

    public void removeEffect(BattleObject battleObject){
        ((Tower)battleObject).removeDamage(battleObjectId);
    }
}
